models/repos.js

```diff
@@ -1,109 +1,205 @@
+    var dblocation = path.join(process.env.HOME)
+    var dbfile = path.join(dblocation, '.dat-desktop-db.json')
+
+    var arr = [
+      function (next) {
+        mkdirp(dblocation, next)
+      },
+      function (val, next) {
+        var db = toilet(dbfile)
+        Multidat(db, next)
+      },
+      function (val, next) {
+        send('repos:ready', function (err) {
+          if (err) return next(err)
+          next(null, val)
+        })
+      }
+    ]
+
+    waterfall(arr, function (err, multidat) {
+      if (err) return done(err)
+      manager = createManager(multidat, function (err, dats) {
+        if (err) return done(err)
+        send('repos:update', dats, done)
+      })
+    })
+  }
+
+  // signal multidat is ready to accept values
+  function multidatReady (state, data) {
+    return { ready: true }
   }
 
-  function update (state, data) {
-    return {
-      updateIndex: state.updateIndex + 1,
-      values: data
-    }
+  // update the values with then new values from the manager
+  function updateDats (state, data) {
+    return { values: data }
   }
 
-  function openDat (state, dat) {
-    open(dat.dir, (err) => {
-      if (err) throw err
+  // handle IPC events from the server
+  function handleIpc (send, done) {
+    ipc.on('link', (event, url) => {
+      const key = encoding.decode(url)
+      send('repos:clone', key, done)
     })
+    ipc.on('log', (ev, str) => console.log(str))
+    ipc.send('ready')
   }
 
+  // open the dat archive in the native filesystem explorer
+  function openDirectory (state, data, send, done) {
+    assert.ok(data.path, 'repos-model.openDirectory: data.path should exist')
+    open(data.path, done)
+  }
+
+  // choose a directory and convert it to a dat archive
   function createDat (state, data, send, done) {
-    const files = dialog.showOpenDialog({
-      properties: ['openDirectory']
+    var pathname = data
+    if (!pathname) {
+      var files = dialog.showOpenDialog({
+        properties: ['openDirectory']
+      })
+      if (!files || !files.length) return done()
+      pathname = files[0]
+    }
+    manager.create(pathname, done)
+  }
+
+  function removeDat (state, data, send, done) {
+    assert.ok(data.key, 'repos-model.deleteDat: data.key should exist')
+
+    if (data.confirmed === true) {
+      const key = encoding.decode(data.key)
+      manager.close(key, done)
+    } else {
+      const encodedKey = encoding.encode(data.key)
+      send('location:set', `?delete=${encodedKey}`, done)
+    }
+  }
+
+  // copy a dat share link to clipboard and open a modal
+  function shareDat (state, data, send, done) {
+    assert.ok(data.key, 'repos-model.shareDat: data.key should exist')
+    const encodedKey = encoding.encode(data.key)
+    send('location:set', `?share=${encodedKey}`, done)
+  }
+
+  function cloneDat (state, data, send, done) {
+    assert.ok(Buffer.isBuffer(data) || typeof data === 'string', 'repos-model.cloneDat: data should be a buffer or a string')
```

Comment:
(hope you ok with this? - I know I use assert a lot, but feel we're still at a spot where we regularly get weird errors that having lots of asserts can help us catch and trace bugs better)